### PR TITLE
feat(Inventory) First draft of a working autocomplete on products and…

### PR DESCRIPTION
**Pull request by Luke1982**
… service lines.

I spent an afternoon getting this in, mostly built on my idlines project. Do have to thoroughly test it, but I tried to keep as much of the original intact. Please test it and let me know what you think. I do need to prevent the annoying behaviour of the Enter key when an autocomplete is open.

### CODE REVIEW

**Comment by Luke1982** [comment]
Also, you need to be able to search on mfr and vendor part no.

File: Smarty/templates/Inventory/ProductDetails.tpl [hunk 1]
```diff
@@ -146,14 +146,25 @@ function displayCoords(currObj,obj,mode,curr_row)
...
+							</svg>
+							</span>
+						</div>
+					</div>
+				</div>
 				<input type="hidden" id="hdnProductId1" name="hdnProductId1" value="{if isset($PRODUCT_ID)}{$PRODUCT_ID}{/if}" />
 				{if $PRODUCT_OR_SERVICE eq 'Services'}
 					<input type="hidden" id="lineItemType1" name="lineItemType1" value="Services" />
-					&nbsp;<img id="searchIcon1" title="{'Services'|@getTranslatedString:'Services'}" src="{'services.gif'|@vtiger_imageurl:$THEME}" style="cursor: pointer;" align="absmiddle" onclick="servicePickList(this,'{$MODULE}',1)" />
+					{*&nbsp;<img id="searchIcon1" title="{'Services'|@getTranslatedString:'Services'}" src="{'services.gif'|@vtiger_imageurl:$THEME}" style="cursor: pointer;" align="absmiddle" onclick="servicePickList(this,'{$MODULE}',1)" />*}
```
Comment (joebordes) [inline Smarty/templates/Inventory/ProductDetails.tpl]: I would not eliminate these buttons as the popup search is more powerful than the autocomplete and you can even create the product/service

File: Smarty/templates/Inventory/ProductDetails.tpl [hunk 2]
```diff
@@ -146,14 +146,25 @@ function displayCoords(currObj,obj,mode,curr_row)
...
 				{else}
 					<input type="hidden" id="lineItemType1" name="lineItemType1" value="Products" />
-					&nbsp;<img id="searchIcon1" title="{'Products'|@getTranslatedString:'Products'}" src="{'products.gif'|@vtiger_imageurl:$THEME}" style="cursor: pointer;" align="absmiddle" onclick="productPickList(this,'{$MODULE}',1)" />
+					{*&nbsp;<img id="searchIcon1" title="{'Products'|@getTranslatedString:'Products'}" src="{'products.gif'|@vtiger_imageurl:$THEME}" style="cursor: pointer;" align="absmiddle" onclick="productPickList(this,'{$MODULE}',1)" />*}
```
Comment (joebordes) [inline Smarty/templates/Inventory/ProductDetails.tpl]: idem as above

**Comment by joebordes** [comment]
I am getting this warning for Services:

```
Warning: number_format() expects parameter 1 to be float, string given in /var/www/coreBOSwork/include/Webservices/CustomerPortalWS.php on line 932
```

File: Smarty/templates/Inventory/ProductDetailsEditView.tpl [hunk 3]
```diff
@@ -190,15 +190,20 @@ function displayCoords(currObj,obj,mode,curr_row)
...
-					<input type="text" id="{$productName}" name="{$productName}" value="{$data.$productName}" class="small" style="width: 70%;" readonly />
+					<div class="slds-combobox_container slds-has-inline-listbox cbds-product-search">
+						<div class="slds-combobox slds-dropdown-trigger slds-dropdown-trigger_click slds-combobox-lookup" aria-expanded="false" aria-haspopup="listbox" role="combobox">
+							<div class="slds-combobox__form-element slds-input-has-icon slds-input-has-icon_right" role="none">
+								<input id="{$productName}" name="{$productName}" class="slds-input slds-combobox__input cbds-inventoryline__input--name" aria-autocomplete="list" aria-controls="listbox-unique-id" autocomplete="off" role="textbox" placeholder="{$APP.typetosearch_prodser}" value="{$data.$productName}" type="text" style="box-shadow: none;">
+								<span class="slds-icon_container slds-icon-utility-search slds-input__icon slds-input__icon_right">
+								<svg class="slds-icon slds-icon slds-icon_x-small slds-icon-text-default" aria-hidden="true">
+									<use xlink:href="lib/LDS/icons/utility-sprite/svg/symbols.svg#search" xmlns:xlink="http://www.w3.org/1999/xlink"></use>
```
Comment (joebordes) [inline Smarty/templates/Inventory/ProductDetailsEditView.tpl]: include/LD/assets/icons/utility-sprite/svg/symbols.svg

File: Smarty/templates/Inventory/ProductDetailsEditView.tpl [hunk 4]
```diff
@@ -190,15 +190,20 @@ function displayCoords(currObj,obj,mode,curr_row)
...
+								</svg>
+								</span>
+							</div>
+						</div>
+					</div>
 					<input type="hidden" id="{$hdnProductId}" name="{$hdnProductId}" value="{$data.$hdnProductId}" />
 					<input type="hidden" id="lineItemType{$row_no}" name="lineItemType{$row_no}" value="{$entityType}" />
-					&nbsp;
```
Comment (joebordes) [inline Smarty/templates/Inventory/ProductDetailsEditView.tpl]: do not eliminate this functionality

File: include/Webservices/CustomerPortalWS.php [hunk 5]
```diff
@@ -864,6 +864,85 @@ function getReferenceAutocomplete($term, $filter, $searchinmodules, $limit, $use
...
 }
 
+/**
+ * @param String $term: search term
+ * @param Array $returnfields: array of fields to return as result, maybe for the future
+ * @param Number $limit: maximum number of values to return
+ * @return Array values found
+ */
+function getProductServiceAutocomplete($term, $returnfields = array(), $limit = 5) {
+	global $adb, $current_user;
```
Comment (joebordes) [inline include/Webservices/CustomerPortalWS.php]: eliminate current_user, you are not using it

File: include/Webservices/CustomerPortalWS.php [hunk 6]
```diff
@@ -864,6 +864,85 @@ function getReferenceAutocomplete($term, $filter, $searchinmodules, $limit, $use
...
+			'name' => $prodser['name'],
+			'divisible' => $prodser['divisible'],
+			'comments' => $prodser['description'],
+			'ven_no' => $prodser['ven_no'],
+			'mfr_no' => $prodser['mfr_no'],
+			'type' => $prodser['type'],
+			'id' => $prodser['id'],
+		);
+		$ret_prodser['pricing'] = array(
+			'unit_price' => number_format($prodser['unit_price'], 2, '.', ''),
```
Comment (joebordes) [inline include/Webservices/CustomerPortalWS.php]: These are the lines that trigger the warning, probably the unit_price/cost are empty or something??

**Comment by Luke1982** [comment]
Basically all your points have been dealt with. Also, the images are back and change dynamically based on if you select a product or service. The onclick handler also changes accordingly. I've setup the autocomplete function to use the current user's no. of decimals.

**Comment by joebordes** [comment]
closing in favor of #333

**Pull request closed without merging**